controllers/humiocluster_controller.go

```diff
@@ -2074,10 +2079,170 @@ func (r *HumioClusterReconciler) ensurePodsExist(ctx context.Context, hc *humiov
+//  5. If the eviction has started, it periodically checks every 60 seconds if the eviction has been completed
+//  6. When the eviction is completed and there is no more data on that node, the node is unregistered from the cluster, and the pod is removed.
+func (r *HumioClusterReconciler) removePodGracefully(ctx context.Context, hc *humiov1alpha1.HumioCluster, podsInNodePool []corev1.Pod, req ctrl.Request) error {
+	// GetCluster gql query returns node ID and Zone
+	clusterConfig, err := helpers.NewCluster(ctx, r, hc.Name, "", hc.Namespace, helpers.UseCertManager(), true, false)
+	if err != nil {
+		return r.logErrorAndReturn(err, fmt.Sprintf("could not create a cluster config for the http client."))
+	}
+	humioHttpClient := r.HumioClient.GetHumioHttpClient(clusterConfig.Config(), req)
+	cluster, err := r.HumioClient.GetCluster(ctx, humioHttpClient, req)
+	if err != nil {
+		return r.logErrorAndReturn(err, "failed to get humio cluster through the GraphQL API")
+	}
+	getCluster := cluster.GetCluster()
+	nodeIdToPodMap := r.matchPodsToHosts(podsInNodePool, getCluster.GetNodes())
+
+	// Check Node Zones and gets the one with the most nodes. In case of a tie, the first zone is used
+	podRemovalZone, err := r.getZoneForPodRemoval(ctx, podsInNodePool)
+	if err != nil {
+		return r.logErrorAndReturn(err, "failed to get pod removal zone")
+	}
+	for _, pod := range podsInNodePool {
+		if pod.Labels[kubernetes.TopologyKubernetesZone] != podRemovalZone {
+			continue
+		}
+		if pod.Spec.NodeName == "" {
+			r.Log.Info(fmt.Sprintf("NodeName is empty for pod %s.", pod.Name))
+			continue
+		}
+		vhost := nodeIdToPodMap[pod.GetName()]
+		err = r.HumioClient.SetIsBeingEvicted(ctx, humioHttpClient, req, vhost, true)
+		if err != nil {
+			return r.logErrorAndReturn(err, fmt.Sprintf("failed to set eviction for vhost %d", vhost))
+		}
+		// wait for eviction status to be updated
+		isEvicted := false
+		for i := 0; i < waitForPodTimeoutSeconds && !isEvicted; i++ {
+			nodesStatus, err := r.getClusterNodesStatus(ctx, humioHttpClient, req)
+			if err != nil {
+				return err
+			}
+
+			for _, node := range nodesStatus {
+				if node.GetId() == vhost && *node.GetIsBeingEvicted() {
+					isEvicted = true
+					break
+				}
+			}
+
+			r.Log.Info(fmt.Sprintf("validating node data eviction is in progress for vhost %d", vhost))
+			time.Sleep(time.Second * 1)
+		}
+
+		if !isEvicted {
+			return r.logErrorAndReturn(err, fmt.Sprintf("failed to set eviction status for vhost %d", vhost))
+		}
+
+		pod.Annotations[kubernetes.PodMarkedForDataEviction] = "true"
+
+		// poll eviction status
+		pollTick := time.Tick(60 * time.Second)
+		for _ = range pollTick {
+			nodesStatus, err := r.getClusterNodesStatus(ctx, humioHttpClient, req)
+			if err != nil {
+				return err
+			}
+			readyForRemoval := false
+			for _, node := range nodesStatus {
+				if node.GetId() == vhost {
+					evictionStatus := node.GetEvictionStatus()
+					if evictionStatus.GetTotalSegmentBytes() == 0 {
```

Comment:
Do we actually have to wait for all bytes to be evicted? Or should we wait until all the data exists on at least one other node?

Reply:
This is outdated ^^, we're currently checking using the `reasonsNodeCannotBeSafelyUnregistered` return field, which provides boolean checks